src/github.com/matrix-org/dendrite/roomserver/input/membership.go

```diff
@@ -0,0 +1,223 @@
+	updates []api.OutputEvent,
+) ([]api.OutputEvent, error) {
+	switch newMembership {
+	case "invite":
+		needsSending, err := mu.SetToInvite(*add)
+		if err != nil {
+			return nil, err
+		}
+		if needsSending {
+			updates = appendInviteUpdate(updates, add)
+		}
+		return updates, nil
+	case "join":
+		if mu.IsJoin() {
+			return updates, nil
+		}
+		retired, err := mu.SetToJoin(add.Sender())
+		if err != nil {
+			return nil, err
+		}
+		return appendRetireUpdates(updates, add, newMembership, retired), nil
+	case "leave", "ban":
+		if mu.IsLeave() {
+			return updates, nil
+		}
+		retired, err := mu.SetToLeave(add.Sender())
+		if err != nil {
+			mu.SetToLeave(add.Sender())
+		}
+		return appendRetireUpdates(updates, add, newMembership, retired), nil
+	default:
+		panic(fmt.Errorf(
+			"input: membership %q is not one of the allowed values", newMembership,
+		))
+	}
+}
+
+func appendInviteUpdate(
+	updates []api.OutputEvent, add *gomatrixserverlib.Event,
+) []api.OutputEvent {
+	onie := api.OutputNewInviteEvent{
+		Event: *add,
+	}
+	return append(updates, api.OutputEvent{
+		Type:           api.OutputTypeNewInviteEvent,
+		NewInviteEvent: &onie,
+	})
+}
+
+func appendRetireUpdates(
+	updates []api.OutputEvent, add *gomatrixserverlib.Event, membership string, retired []string,
+) []api.OutputEvent {
+	for _, eventID := range retired {
+		orie := api.OutputRetireInviteEvent{
+			EventID:    eventID,
+			Membership: membership,
+		}
+		if add != nil {
+			orie.RetiredByEventID = add.EventID()
+		}
+		updates = append(updates, api.OutputEvent{
+			Type:              api.OutputTypeRetireInviteEvent,
+			RetireInviteEvent: &orie,
+		})
+	}
+	return updates
+}
+
+type stateChange struct {
+	removed types.StateEntry
+	added   types.StateEntry
+}
+
+func pairUpChanges(removed, added []types.StateEntry) []stateChange {
+	var ai int
+	var ri int
+	var result []stateChange
+	for {
+		switch {
+		case ai == len(added):
+			for _, s := range removed[ri:] {
+				result = append(result, stateChange{removed: s})
+			}
+			return result
+		case ri == len(removed):
+			for _, s := range added[ai:] {
+				result = append(result, stateChange{added: s})
+			}
+			return result
+		case added[ai].StateKeyTuple == removed[ri].StateKeyTuple:
+			result = append(result, stateChange{
+				removed: removed[ri],
+				added:   added[ai],
+			})
+			ai++
+			ri++
+		case added[ai].StateKeyTuple.LessThan(removed[ri].StateKeyTuple):
+			result = append(result, stateChange{added: added[ai]})
+			ai++
+		default:
+			result = append(result, stateChange{removed: removed[ri]})
+			ri++
+		}
+	}
+}
+
+// membershipChanges pairs up the membership state changes from a sorted list
+// of state removed and a sorted list of state added.
+func membershipChanges(removed, added []types.StateEntry) []stateChange {
+	changes := pairUpChanges(removed, added)
+	var result []stateChange
+	for _, c := range changes {
+		if c.added.EventTypeNID == types.MRoomMemberNID ||
```

Comment:
It's rather curious you've chosen to design this code like this. So, as I understand it:
 - We pair up all state changes in the removed/added lists.
 - We then filter out all state changes except for the `m.room.member` ones.
 - We return this filtered list of state changes.

Why do we bother working out the list of all state changes if we're just going to discard (most of?) them immediately? We never call `pairUpChanges` anywhere else.